### PR TITLE
MLT-0053 Add use-case for moving items

### PR DESCRIPTION
**Pull request by anotheroneofthese**
Originally planned as updating items - this PR implements the functionality to move items within WLS. This specific use-case is required for when the storage systems gives us various status updates when the items are withdrawn, inserted, or generally moved around

### CODE REVIEW

File: src/main/kotlin/no/nb/mlt/wls/domain/ports/outbound/ItemRepository.kt [hunk 1]
```diff
@@ -13,6 +13,15 @@ interface ItemRepository {
     fun createItem(item: Item): Mono<Item>
 
     suspend fun doesAllItemsExist(ids: List<ItemId>): Boolean
+
```
Comment (MormonJesus69420) [inline src/main/kotlin/no/nb/mlt/wls/domain/ports/outbound/ItemRepository.kt:16]: I know it's unrelated, but can we fix the grammar of the above metod to "doAllItemsExist" or "doesEveryItemExist"?

File: src/main/kotlin/no/nb/mlt/wls/domain/ports/inbound/MoveItem.kt [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+ * Some examples include handling status messages when crates arrive
+ * at picking stations, or when items return to the
+ * storage systems.
+ * In both cases we want to know where the item went, and if the count changed
+ */
+interface MoveItem {
+    suspend fun moveItem(
+        hostId: String,
+        hostName: HostName,
+        quantity: Double,
```
Comment (MormonJesus69420) [inline src/main/kotlin/no/nb/mlt/wls/domain/ports/inbound/MoveItem.kt:18]: Don't know where the best place is to check, but quantity should be zero or positive, would be funky if we have -1 book.

File: src/main/kotlin/no/nb/mlt/wls/domain/ports/inbound/MoveItem.kt [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+        location: String
```
Comment (MormonJesus69420) [inline src/main/kotlin/no/nb/mlt/wls/domain/ports/inbound/MoveItem.kt:19]: Should check if location is not a blank value. Could probably do that and the above check with annotations, or do a logic check somewhere else in the logic flow, probably service if I were to make the call.

Which reminds me that we should also make sure that when an item is registered the location is set to a null, but only then, cause it's the only case where location can be null, at least so far.

**Review by MormonJesus69420 — CHANGES_REQUESTED** [review]
Looks good for now, just a few minor niggles to iron out.

Also, we need to keep in mind that in future we will need to do some more work when item moves.

For example, checking if item moves into the storage system, and is part of an order, so we can update the item as returned. Will have to discuss how to best model and implement that kind of logic.

Other than that, it's looking great!

File: src/main/kotlin/no/nb/mlt/wls/domain/ports/inbound/MoveItem.kt (hunk 3 above)
Comment (MormonJesus69420) [reply src/main/kotlin/no/nb/mlt/wls/domain/ports/inbound/MoveItem.kt:19]: With those changes made, we should also have tests to validate that of course.

Comment (anotheroneofthese) [reply src/main/kotlin/no/nb/mlt/wls/domain/ports/inbound/MoveItem.kt:19]: For the latter: Items are registered using `ItemMetadata`, which when mapped to `Item` sets these fields to null. They will only be mapped if you explictly set them. If you'd like I could try writing a comment which describes this intention

File: src/test/kotlin/no/nb/mlt/wls/domain/WLSServiceTest.kt [hunk 4]
```diff
@@ -114,8 +115,77 @@ class WLSServiceTest {
...
+    }
+
+    @Test
+    fun `moveItem throws when count is invalid`() {
+        coEvery { itemRepoMock.moveItem(any(), any(), -1.0, any()) } throws ValidationException("Location cannot be blank")
+
+        val cut = WLSService(itemRepoMock, orderRepoMock, storageSystemRepoMock)
+        runTest {
+            assertThrows<RuntimeException> {
+                cut.moveItem(testItem.hostId, testItem.hostName, -1.0, "   ")
```
Comment (MormonJesus69420) [inline src/test/kotlin/no/nb/mlt/wls/domain/WLSServiceTest.kt:167]:
```suggestion
                cut.moveItem(testItem.hostId, testItem.hostName, -1.0, "Somewhere nice")
```

**Review by MormonJesus69420 — APPROVED** [review]
Looks great!

**Pull request merged**